numba/_typeof.c

```diff
@@ -831,6 +834,52 @@ int typecode_arrayscalar(PyObject *dispatcher, PyObject* aryscalar) {
     return BASIC_TYPECODES[typecode];
 }
 
+static
+int typecode_devicendarray(PyObject *dispatcher, PyObject *ary)
+{
+    int typecode;
+    int dtype;
+    int ndim;
+    int layout = 0;
+
+    PyObject* flags = PyObject_GetAttrString(ary, "flags");
+    if (PyDict_GetItemString(flags, "C_CONTIGUOUS") == Py_True) {
+        layout = 1;
+    } else if (PyDict_GetItemString(flags, "F_CONTIGUOUS") == Py_True) {
+        layout = 2;
+    }
+
+    ndim = PyLong_AsLong(PyObject_GetAttrString(ary, "ndim"));
+
+    if (ndim <= 0 || ndim > N_NDIM) goto FALLBACK;
+
+    PyObject* dtype_obj = PyObject_GetAttrString(ary, "dtype");
+    int dtype_num = PyLong_AsLong(PyObject_GetAttrString(dtype_obj, "num"));
+    dtype = dtype_num_to_typecode(dtype_num);
+    if (dtype == -1) goto FALLBACK;
+
+    /* Fast path, using direct table lookup */
+    assert(layout < N_LAYOUT);
+    assert(ndim <= N_NDIM);
```

Comment:
whilst not strictly necessary as I'm not sure how you'd created one in CUDA, perhaps also add this just to be on the safe side
```suggestion
    assert(ndim <= N_NDIM);
    assert(dtype < N_DTYPES);
```

Reply:
Added.